File: leetcode/js/15.3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    /// 三指针
    nums.sort((a, b) => a - b);
    let res = []
    for (let i = 0; i < nums.length - 2; ++i) {
        if (i !== 0) {
            if (nums[i] === nums[i - 1]) continue;//重复的元素
        }
        let left = i + 1, right = nums.length - 1;
        while (left < right) {
            let sum_v = nums[i] + nums[left] + nums[right];
            if (sum_v === 0) {
                res.push([nums[i], nums[left], nums[right]]);
                ++left;
                while (left < right && nums[left] === nums[left - 1])++left;//跳过重复的元素
                --right;
                while (right > left && nums[right] === nums[right + 1])--right; //跳过重复的元素

            }
            else {
                if (sum_v < 0) {
                    ++left;
                }
                else {
                    --right;
                }
            }
        }
    }
    return res;
};

File: leetcode/scala/15.3sum.scala
import scala.collection.mutable.ListBuffer

object Solution {
  def threeSum(nums: Array[Int]): List[List[Int]] = {
    val arr: Array[Int] = nums.sorted
    val res = ListBuffer[List[Int]]()
    for (i <- arr.indices) {
      if (!(i > 0 && arr(i) == arr(i - 1))) {
        var l = i + 1
        var r = arr.length - 1
        val n = arr(i)

        while (l < r) {
          val sum_ = arr(l) + arr(r)
          if (sum_ + n == 0) {
            res.append(List[Int](arr(i), arr(l), arr(r)))
            l += 1

            while (l < arr.length && arr(l) == arr(l - 1))
              l += 1
            r -= 1
            while (r >= 0 && arr(r) == arr(r + 1))
              r -= 1

          } else {
            if (sum_ + n > 0) {
              r -= 1
            }
            else {
              l += 1
            }
          }
        }
      }
    }

    res.toList
  }
}